Fixes/Preparation for json deserializer

matthiasbeyer opened the pull request:
This is another rip from #260 which fixes a bunch of things. #260 itself needs
to be rebased on this and could then be merged, so please review carefully.

matthiasbeyer committed:
Fix: ws_transaction_push_statement()

This patch fixes the ws_transaction_push_statement() function by
doing a (almost) complete rewrite.

The code was tested in another branch.

matthiasbeyer committed:
Fix: ws_transaction_new() comparsion

matthiasbeyer committed:
Fix: deserializer_state(_new())

This patch fixes the deserializer_state type as well as the
deserializer_state_new() function. We don't need to cache the allocation
type here. It also adds an initialization for the bracket counting and
the ID (zero is an invalid ID now).

neithernut commented on src/command/statement.c:
```diff
@@ -163,6 +163,8 @@ command_args_append(
-    return newargs + self->num++;
+    struct ws_argument* ret = &newargs[self->num];
+    self->num++;
+    return ret;
```
Please describe **exactely** what problems you did have with the previous solution, because from a functional view, there is **no** change introduced by this commit --at least no change which **I** would be aware of.

If you fail to justify this changeset, I'll refuse to ACK this commit.
If you manage to justify this changeset, there is one termination point in the same function which may also be malicious.

---

Off topic: obviously, our sense in aestetics differs a great deal here.

matthiasbeyer replied:
I just re-run the tests without this patch and they still succeed, so it was another patch which fixed my problems. I will remove this patch from the changeset now.

neithernut commented on src/serialize/json/deserializer_state.c:
```diff
@@ -55,9 +55,13 @@ deserialize_state_new(
     state->tmp_statement    = NULL;
 
+    state->nboxbrackets     = 0;
+    state->ncurvedbrackets  = 0;
```
You _could_ use one integer for both maps and arrays...
But this would be a rather aestetical (and thus maybe unnecessary) change.

neithernut commented:
Woops, mis-clicked.

neithernut commented on src/command/command.h:
```diff
@@ -60,7 +60,7 @@ typedef int (*ws_special_command_func)(struct ws_processor*,
  */
 struct ws_command {
-    char const* const name; //!< @public name of the command
+    char const* name; //!< @public name of the command
```
I think this changeset is a result of bad usage.

You can omit this changeset now, since you don't even have anything to do with a command anymore (you sould use `ws_statement_init()` anyways). You shouldn't even include the `command.h` header.

matthiasbeyer commented:
Fixed what you said (btw I'm keeping #260 up to date and I remove the patches from there when removing them here. When rebasing, the most stuff will be thrown out automatically, but not the patches I removed here, so I'm doing this by hand every time, so I don't have to do this afterwards).

neithernut commented:
ACK.
